main.py

```diff
@@ -21,7 +21,60 @@ def parse(self, response):
         # open_in_browser(response)
 
     def parseResults(self, response):
-        test = response.xpath("//div[@id='ctl00_PlaceHolderMain_RecordSearchResultInfo_noDataMessageForSearchResultList_messageBar']")
-        test2 = response.xpath("//table[@id='ctl00_PlaceHolderMain_dgvPermitList_gdvPermitList']")
-        print("RESULTS", test, test2)
-        open_in_browser(response)
+        results_empty = response.xpath("//div[@id='ctl00_PlaceHolderMain_RecordSearchResultInfo_noDataMessageForSearchResultList_messageBar']").get()
+        records_rows = response.xpath("//table[@id='ctl00_PlaceHolderMain_dgvPermitList_gdvPermitList']/tr")
+
+        # Make sure that the page format is as expected.
+        # we expect that there's either no records with a message explaining that, or there are records without a message.
+        # If that's not true, it's likely that the page format has changed and this spider needs to be updated.
+        if results_empty:
+            assert len(records_rows) == 0, "Expected no records, but records were found"
+        else:
+            assert len(records_rows) > 0, "Expected records, but none were found"
+
+        # if there are records, extract them
+        if records_rows:
+            records = self.extractRecords(response, records_rows)
+            for record in records:
+                print(record)
+
+
+    def extractRecords(self, response, records_rows):
+        records = []
+        for row in records_rows:
+            # if this row is a header row, skip it
+            if row.xpath(".//th"):
+                continue
+
+            # if this row is the pagination row, skip it
+            if row.xpath(".//table"):
+                continue
+
+            # select the data we want from the table row
+            record_number = row.xpath("string(.//td[2])").get().strip()
+            record_link = row.xpath(".//td[2]//a/@href").get()
+            record_type = row.xpath("string(.//td[3])").get().strip()
+            project_name = row.xpath("string(.//td[4])").get().strip()
+            address = row.xpath("string(.//td[5])").get().strip()
+
+            # make sure that the data is as expected
+            assert record_number, "Record number not found"
+            assert record_link, "Record link not found"
+            assert record_type, "Record type not found"
+            assert address, "Address not found"
```

Comment:
I don't think that the `assert`s are going to help you out here, `xpath` will throw an `AssertionError` if you try, say,

```
address = row.xpath("string(.//td[7])").get()
``` 

IMHO better solution would be try/except or testing `row.xpath("whatever")` before calling `.get()`.

Since we're still building out this script I don't think it has to be super hardened, but those `assert`s may be misleading.

Reply:
The intent with those is also to check that the value string actually has content, as opposed to the project_name for instance which seems to be allowed to be an empty string (based on looking at real results).

At least with the scrapy version i'm using when I try faking an xpath selection that shouldn't work, like `address = row.xpath("string(.//td[7])").get()` the bug also gets caught by these assertions. However, I agree if scrapy has a way to throw when the xpath result is empty it would give a more specific error and would be a plus.

Reply:
I haven't tried them, but there are [selectors in xpath syntax](https://stackoverflow.com/questions/15909348/how-to-tell-using-xpath-if-an-element-is-present-and-non-empty) that can be used to check for a "exists but empty" status. I don't believe Scrapy has anything specific for that.

I appreciate the thoroughness, but I agree with @thashepherd - I wouldn't rely on `asserts` here. Try/except would be better. We could also recover from those errors and potentially include those errors in the email, which could help keep us from failing silently.

Reply:
Makes sense, I just removed them in the new commit